### ISSUE TITLE
Segfault on Linux when changing mouse cursor

### ISSUE THREAD

**Issue by tommag** [juce-framework/JUCE#720]
Hello, 

I have observed on Linux 64 bits (ArchLinux / Gnome / X11) a crash that looks related to this commit : https://github.com/juce-framework/JUCE/commit/8e4277085aad6c2495523a6068f373a5c672558c

It happens on this piece of software : https://github.com/benkuper/Chataigne (only on Linux) when hovering a ResizableCornerComponent. It happens not on a fresh start, but after loading a file (which means that the UI is destroyed and recreated).

I tried to include as much information as possible below as I believe that there is an unexpected behavior in `juce_gui_basics/native/juce_linux_X11_Windowing.cpp` (the same code worked fine before the commit I mentioned above, and it only involves JUCE components). 

Debug session screenshot just before the segfault:
![Capture d’écran de 2020-05-13 18-57-14](https://user-images.githubusercontent.com/5353407/81845948-a026ac00-9551-11ea-911f-b8e4ef0292bc.png)
Segfault:
![Capture d’écran de 2020-05-13 18-58-17](https://user-images.githubusercontent.com/5353407/81846022-baf92080-9551-11ea-8a40-dd0c47af0311.png)

Full stack trace : 

> libX11.so.6!XFreeCursor (Unknown Source:0)
juce::MouseCursor::showInWindow(const juce::MouseCursor * const this, juce::ComponentPeer * peer) (/home/tom/JUCE/modules/juce_gui_basics/native/juce_linux_X11_Windowing.cpp:4125)
juce::MouseInputSourceInternal::showMouseCursor(juce::MouseInputSourceInternal * const this, juce::MouseCursor cursor, bool forcedUpdate) (/home/tom/JUCE/modules/juce_gui_basics/mouse/juce_MouseInputSource.cpp:490)
juce::MouseInputSourceInternal::revealCursor(juce::MouseInputSourceInternal * const this, bool forcedUpdate) (/home/tom/JUCE/modules/juce_gui_basics/mouse/juce_MouseInputSource.cpp:506)
juce::MouseInputSourceInternal::setComponentUnderMouse(juce::MouseInputSourceInternal * const this, juce::Component * newComponent, juce::Point<float> screenPos, juce::Time time) (/home/tom/JUCE/modules/juce_gui_basics/mouse/juce_MouseInputSource.cpp:252)
juce::MouseInputSourceInternal::setScreenPos(juce::MouseInputSourceInternal * const this, juce::Point<float> newScreenPos, juce::Time time, bool forceUpdate) (/home/tom/JUCE/modules/juce_gui_basics/mouse/juce_MouseInputSource.cpp:270)
juce::MouseInputSourceInternal::handleEvent(juce::MouseInputSourceInternal * const this, juce::ComponentPeer & newPeer, juce::Point<float> positionWithinPeer, juce::Time time, const juce::ModifierKeys newMods, float newPressure, float newOrientation, juce::PenDetails pen) (/home/tom/JUCE/modules/juce_gui_basics/mouse/juce_MouseInputSource.cpp:340)
juce::MouseInputSource::handleEvent(juce::MouseInputSource * const this, juce::ComponentPeer & peer, juce::Point<float> pos, juce::int64 time, juce::ModifierKeys mods, float pressure, float orientation, const juce::PenDetails & penDetails) (/home/tom/JUCE/modules/juce_gui_basics/mouse/juce_MouseInputSource.cpp:634)
juce::ComponentPeer::handleMouseEvent(juce::ComponentPeer * const this, juce::MouseInputSource::InputSourceType type, juce::Point<float> pos, juce::ModifierKeys newMods, float newPressure, float newOrientation, juce::int64 time, juce::PenDetails pen, int touchIndex) (/home/tom/JUCE/modules/juce_gui_basics/windows/juce_ComponentPeer.cpp:88)
juce::LinuxComponentPeer::handleMotionNotifyEvent(juce::LinuxComponentPeer * const this, const XPointerMovedEvent & movedEvent) (/home/tom/JUCE/modules/juce_gui_basics/native/juce_linux_X11_Windowing.cpp:1926)
juce::LinuxComponentPeer::handleWindowMessage(juce::LinuxComponentPeer * const this, XEvent & event) (/home/tom/JUCE/modules/juce_gui_basics/native/juce_linux_X11_Windowing.cpp:1638)
juce::WindowingHelpers::windowMessageReceive(XEvent & event) (/home/tom/JUCE/modules/juce_gui_basics/native/juce_linux_X11_Windowing.cpp:3477)
juce::XWindowSystem::<lambda(int)>::operator()(int) const(const juce::XWindowSystem::<lambda(int)> * const __closure) (/home/tom/JUCE/modules/juce_gui_basics/native/juce_linux_X11.cpp:217)
std::_Function_handler<void(int), juce::XWindowSystem::initialiseXDisplay()::<lambda(int)> >::_M_invoke(const std::_Any_data &, int &&)(const std::_Any_data & __functor,  __args#0) (/usr/include/c++/9.3.0/bits/std_function.h:300)
std::function<void (int)>::operator()(int) const(const std::function<void(int)> * const this,  __args#0) (/usr/include/c++/9.3.0/bits/std_function.h:688)
juce::InternalRunLoop::dispatchPendingEvents(juce::InternalRunLoop * const this) (/home/tom/JUCE/modules/juce_events/native/juce_linux_Messaging.cpp:186)
juce::MessageManager::dispatchNextMessageOnSystemQueue(bool returnIfNoPendingMessages) (/home/tom/JUCE/modules/juce_events/native/juce_linux_Messaging.cpp:294)
juce::MessageManager::runDispatchLoop(juce::MessageManager * const this) (/home/tom/JUCE/modules/juce_events/messages/juce_MessageManager.cpp:128)
juce::JUCEApplicationBase::main() (/home/tom/JUCE/modules/juce_events/messages/juce_ApplicationBase.cpp:262)
juce::JUCEApplicationBase::main(int argc, const char ** argv) (/home/tom/JUCE/modules/juce_events/messages/juce_ApplicationBase.cpp:240)

As you can see, the crash is caused by the cursorMap incorrectly filled (references cursor 33554440 with display 0x0, then trying to call XFreeCursor with this incorrect information). 

The really strange part is that I have tried to find when this incorrect cursorMap entry is created, and even with placing breakpoints on all cursorMap accesses, I couldn't see it, nor the `[0]: 0x0` entry creation. However all of the correct entries creations and deletions are captured by the breakpoints.

Let me know if I can provide any more information.

Thanks !

**Comment by tommag** [comment]
I took some time to dig a little further and here are my findings : 

- The root cause of this crash is that on window load, the first two cursor creations (calling `XCreateFontCursor` with the same display, but with shape=14 for BottomRightCornerResizeCursor then with shape=116 for UpDownResizeCursor) return the same cursor handle. 

I have not (yet) managed to reproduce this in a minimal example and can't really understand where it comes from.
The consequence of this is that the `cursorMap` has only one entry for two cursors. 

- Then after a GUI reload (opening a previously saved file), all of the GUI component is destroyed and re-created. Thus, all of the cursors are freed and the `cursorMap` is empty. The catch is that, as the window is resizable, the BottomRightCornerResizeCursor is not destroyed ; however it has no associated entry in the `cursorMap`, as it had initially the same cursor handle as the UpDownResizeCursor.

- The next time that the BottomRightCornerResizeCursor is needed inside the GUI (it works fine to resize the window), as there is no entry for it in the `cursorMap`, an empty entry is created at this line : https://github.com/juce-framework/JUCE/commit/8e4277085aad6c2495523a6068f373a5c672558c#diff-de163a7a9419a3f00ee482ac59be861dR4120
which causes a segfault when calling `XFreeCursor` with a null display (see my original screenshot).

(as an aside, the `cursorMap`  `[0]: 0x0` entry is also created at this line, when the cursor handle is null).



While researching this issue, I've also found this post : https://bugs.freedesktop.org/show_bug.cgi?id=9248 which suggests that calling `XFreeCursor` is not needed anyway when the display is invalid. 

For now, I'm commenting that `XFreeCursor` statement :)
